extension/server/extension.go

```diff
@@ -0,0 +1,163 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/jellydator/ttlcache/v3"
+	jsoniter "github.com/json-iterator/go"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/extension"
+	"go.uber.org/zap"
+
+	"github.com/aws/amazon-cloudwatch-agent/extension/entitystore"
+	tlsInternal "github.com/aws/amazon-cloudwatch-agent/internal/tls"
+)
+
+type Server struct {
+	logger         *zap.Logger
+	config         *Config
+	jsonMarshaller jsoniter.API
+	httpsServer    *http.Server
+	ctx            context.Context
+	watcher        *tlsInternal.CertWatcher
+}
+
+var _ extension.Extension = (*Server)(nil)
+
+func (s *Server) setRouter(router *gin.Engine) {
+	router.Use(gin.Recovery())
+	//disabling the gin default behavior of encoding/decoding the request path
+	router.UseRawPath = true
+	router.UnescapePathValues = false
+	router.GET("/kubernetes/pod-to-service-env-map", s.k8sPodToServiceMapHandler)
+}
+
+func NewServer(logger *zap.Logger, config *Config) *Server {
+	s := &Server{
+		logger:         logger,
+		config:         config,
+		jsonMarshaller: jsoniter.ConfigCompatibleWithStandardLibrary,
+		ctx:            context.Background(),
+	}
+	gin.SetMode(gin.ReleaseMode)
+
+	// Initialize a new cert watcher with cert/key pair
+	watcher, err := tlsInternal.NewCertWatcher(config.TLSCertPath, config.TLSKeyPath, config.TLSCAPath, logger)
+	if err != nil {
+		s.logger.Error("failed to initialize cert watcher", zap.Error(err))
+		return s
+	}
+
+	s.watcher = watcher
+
+	watcher.RegisterCallback(func() {
+		s.logger.Debug("Calling registered callback, reloading TLS server")
+		if err := s.reloadServer(watcher.GetTLSConfig()); err != nil {
+			s.logger.Error("Failed to reload TLS server", zap.Error(err))
+		}
+	})
+
+	// Start goroutine with certwatcher running fsnotify against supplied certdir
+	go func() {
+		if err := watcher.Start(s.ctx); err != nil {
+			s.logger.Error("failed to start cert watcher", zap.Error(err))
```

Comment:
So it logs that the watcher couldn't start and then exits the goroutine, but the server just keeps going without TLS. Is that the right behavior?

Reply:
Agreed to keep it this way as restarting the agent pod on server extension start up failure might be too aggressive.